File: leetCode-swift/leetCode-swift/leetCode-0005.swift
//
//  leetCode-0005.swift
//  leetCode-swift
//
//  Created by liyoucheng on 2018/11/21.
//  Copyright © 2018年 Giga. All rights reserved.
//

import Foundation

extension String {
    func substring(from: Int?, to: Int?) -> String {
        if let start = from {
            guard start < self.count else {
                return ""
            }
        }
        
        if let end = to {
            guard end >= 0 else {
                return ""
            }
        }
        
        if let start = from, let end = to {
            guard end - start >= 0 else {
                return ""
            }
        }
        
        let startIndex: String.Index
        if let start = from, start >= 0 {
            startIndex = self.index(self.startIndex, offsetBy: start)
        } else {
            startIndex = self.startIndex
        }
        
        let endIndex: String.Index
        if let end = to, end >= 0, end < self.count {
            endIndex = self.index(self.startIndex, offsetBy: end + 1)
        } else {
            endIndex = self.endIndex
        }
        
        return String(self[startIndex ..< endIndex])
    }
    
    func substring(from: Int?, length: Int) -> String {
        guard length > 0 else {
            return ""
        }
        
        let end: Int
        if let start = from, start > 0 {
            end = start + length - 1
        } else {
            end = length - 1
        }
        
        return self.substring(from: from, to: end)
    }
}

func longestPalindrome(_ s: String) -> String {

    var t: [Character] = []
    t.append("$")
    t.append("#")
    for c in s {
        t.append(c)
        t.append("#")
    }
    
    var p = [Int](repeatElement(0, count: t.count))
    var mx = 0
    var id = 0
    var resLen = 0
    var resCenter = 0
    for i in t.indices {
        if mx > i {
            p[i] = min(p[2 * id - i], mx - i)
        }
        else {
            p[i] = 1
        }
        
        
        while ((i >= p[i]) && (i + p[i] < t.count) && (t[i + p[i]] == t[i - p[i]])) {
            p[i] += 1
        }

        if (mx < i + p[i]) {
            mx = i + p[i]
            id = i
        }
        
        if (resLen < p[i]) {
            resLen = p[i]
            resCenter = i
        }
        
    }

    return s.substring(from: (resCenter - resLen) / 2, length: resLen - 1)
}

func test_0005() {
    let s = longestPalindrome("babad")
    print(s)
}

File: leetCode-c/leetCode-c/LeetCode/leetCode-0005/leetCode-0005.c
//
//  leetCode-0005.c
//  leetCode-c
//
//  Created by liyoucheng on 2018/11/21.
//  Copyright © 2018年 Giga. All rights reserved.
//

#include "leetCode-0005.h"

#include <stdlib.h>
#include <string.h>

#include "comman.h"

//最长子串的长度是半径减1，起始位置是中间位置减去半径再除以2。
//p[i] = mx > i ? min(p[2 * id - i], mx - i) : 1;
int manacher(char *s, char output[])
{
    int i, j;
    char t[3000] = { '\0' };
    
    // 填充$和#
    t[0] = '$';
    for (i = 0; s[i] != '\0'; i++) {
        t[(i<<1)+1] = '#';
        t[(i<<1)+2] = s[i];
    }
    t[(i<<1)+1] = '#';
    int len = (i<<1)+2;
    t[len] = '\0';
    
    int p[3000] = { 0 }; // 以t中某一点为中心的回文半径
    int id = 0; // 回文的中心点
    int mx = 0; // 最长回文的右边界点
    int resLen = 0; // 最大回文长度
    int resCenter = 0; // 最长回文的中心点
    for (i = 1; i < len; i++) {
        if (i < mx) {
            p[i] = min(p[2*id-i], mx-i);
        } else {
            p[i] = 1;
        }
        
        while (t[i+p[i]] == t[i-p[i]]) {
            p[i]++;
        }
        
        if (i+p[i] > mx) {
            mx = i+p[i];
            id = i;
        }
        
        if (resLen < p[i]-1) {
            resLen = p[i]-1;
            resCenter = i;
        }
    }
    
    for (j = 0, i = resCenter - resLen; i <= resCenter + resLen; i++) {
        if (t[i] != '#') {
            output[j++] = t[i];
        }
    }
    return resLen;
}

char *longestPalindrome(char *s)
{
    if (s == NULL) {
        return NULL;
    }
    
    int len = (int)strlen(s);
    if (len <= 1) {
        return s;
    }
    
    char *palindrome = malloc(2000);
    memset(palindrome, 0, sizeof(malloc(2000)));
    
    int size = manacher(s, palindrome);
    palindrome[size] = '\0';
    return palindrome;
}

void test_0005() {
    char *s = "babad";
    char *ret = longestPalindrome(s);
    printf("%s\n",ret);
}


